File: shared/main/java/net/ccbluex/liquidbounce/utils/item/ItemUtils.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils.item;

import java.util.Objects;
import java.util.regex.Pattern;
import java.util.stream.IntStream;

import net.ccbluex.liquidbounce.api.minecraft.enchantments.IEnchantment;
import net.ccbluex.liquidbounce.api.minecraft.item.IItem;
import net.ccbluex.liquidbounce.api.minecraft.item.IItemStack;
import net.ccbluex.liquidbounce.api.minecraft.util.IResourceLocation;
import net.ccbluex.liquidbounce.utils.ClientUtils;
import net.ccbluex.liquidbounce.utils.MinecraftInstance;

import org.jetbrains.annotations.Contract;

/**
 * @author MCModding4K
 */
public final class ItemUtils extends MinecraftInstance
{

	/**
	 * Allows you to create a item using the item json
	 *
	 * @param  itemArguments
	 *                       arguments of item
	 * @return               created item
	 * @author               MCModding4K
	 */
	public static IItemStack createItem(String itemArguments)
	{
		try
		{
			itemArguments = itemArguments.replace('&', '\u00A7');
			IItem item = classProvider.createItem();
			final IItem itemInstance = item;
			String[] args = null;

			for (int mode = 0, modeSize = Math.min(12, itemArguments.length() - 2); mode <= modeSize; ++mode)
			{
				args = itemArguments.substring(mode).split(Pattern.quote(" "));
				final IResourceLocation resourcelocation = classProvider.createResourceLocation(args[0]);
				item = functions.getObjectFromItemRegistry(resourcelocation);

				if (item != null)
					break;
			}

			if (item == null)
				return null;

			int i = 1;
			if (args.length >= 2 && PATTERN.matcher(args[1]).matches())
				i = Integer.parseInt(args[1]);

			int j = 0;
			if (args.length >= 3 && PATTERN.matcher(args[2]).matches())
				j = Integer.parseInt(args[2]);

			final IItemStack itemstack = classProvider.createItemStack(item, i, j);

			if (args.length >= 4)
			{
				final StringBuilder NBT = new StringBuilder();
				final int argsLength = args.length;
				for (int nbtcount = 3; nbtcount < argsLength; ++nbtcount)
					NBT.append(" ").append(args[nbtcount]);
				itemstack.setTagCompound(classProvider.getJsonToNBTInstance().getTagFromJson(NBT.toString()));
			}

			return itemstack;
		}
		catch (final Exception e)
		{
			// noinspection StringConcatenationArgumentToLogCall
			ClientUtils.getLogger().error("Can't create the item with arguments \"" + itemArguments + "\"", e);
			return null;
		}
	}

	public static int getEnchantment(final IItemStack itemStack, final IEnchantment enchantment)
	{
		return itemStack == null || itemStack.getEnchantmentTagList() == null || itemStack.getEnchantmentTagList().hasNoTags() ? 0 : IntStream.range(0, itemStack.getEnchantmentTagList().tagCount()).mapToObj(i -> itemStack.getEnchantmentTagList().getCompoundTagAt(i)).filter(tagCompound -> tagCompound.hasKey("ench") && tagCompound.getShort("ench") == enchantment.getEffectId() || tagCompound.hasKey("id") && tagCompound.getShort("id") == enchantment.getEffectId()).findFirst().map(tagCompound -> tagCompound.getShort("lvl")).orElseGet(() -> (short) 0);
	}

	public static int getEnchantmentCount(final IItemStack itemStack)
	{
		return itemStack == null || itemStack.getEnchantmentTagList() == null || itemStack.getEnchantmentTagList().hasNoTags() ? 0 : (int) IntStream.range(0, itemStack.getEnchantmentTagList().tagCount()).mapToObj(i -> itemStack.getEnchantmentTagList().getCompoundTagAt(i)).filter(tagCompound -> tagCompound.hasKey("ench") || tagCompound.hasKey("id")).count();
	}

	@Contract("null -> true")
	public static boolean isStackEmpty(final IItemStack stack)
	{
		return stack == null || classProvider.isItemAir(stack.getItem());
	}

	private static final Pattern PATTERN = Pattern.compile("\\d+"); // TODO: Rename
}

File: shared/main/java/net/ccbluex/liquidbounce/utils/item/ItemUtils.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.utils.item

import net.ccbluex.liquidbounce.api.minecraft.enchantments.IEnchantment
import net.ccbluex.liquidbounce.api.minecraft.item.IItem
import net.ccbluex.liquidbounce.api.minecraft.item.IItemStack
import net.ccbluex.liquidbounce.api.minecraft.nbt.INBTTagCompound
import net.ccbluex.liquidbounce.utils.ClientUtils.logger
import net.ccbluex.liquidbounce.utils.MinecraftInstance
import java.util.regex.Pattern
import kotlin.math.min

/**
 * @author MCModding4K
 */
object ItemUtils : MinecraftInstance()
{
	/**
	 * Allows you to create a item using the item json
	 *
	 * @param  itemArguments
	 * arguments of item
	 * @return               created item
	 * @author               MCModding4K
	 */
	@JvmStatic
	fun createItem(itemArguments: String): IItemStack?
	{
		return try
		{
			val fixedItemArguments = itemArguments.replace('&', '\u00A7') // Translate Colorcodes

			var item: IItem? = classProvider.createItem()
			var args: List<String>? = null

			val modeSize = min(12, fixedItemArguments.length - 2)


			run {
				repeat(modeSize) { i ->
					args = fixedItemArguments.substring(i).split(" ")
					item = functions.getObjectFromItemRegistry(classProvider.createResourceLocation((args ?: return@repeat)[0]))
					if (item != null) return@run
				}
			}

			item ?: return null

			val checkedArgs = args ?: return null
			val argsLength = checkedArgs.size

			// Item amount
			var amount = 1
			if (argsLength >= 2 && PATTERN.matcher(checkedArgs[1]).matches()) amount = checkedArgs[1].toInt()

			// Item meta
			var meta = 0
			if (argsLength >= 3 && PATTERN.matcher(checkedArgs[2]).matches()) meta = checkedArgs[2].toInt()

			val itemstack = classProvider.createItemStack(item!!, amount, meta)

			// Build NBT tag
			if (argsLength >= 4)
			{
				val nbtBuilder = StringBuilder()

				for (nbtcount in 3 until argsLength) nbtBuilder.append(" ${checkedArgs[nbtcount]}")

				itemstack.tagCompound = classProvider.jsonToNBTInstance.getTagFromJson("$nbtBuilder")
			}

			itemstack
		}
		catch (e: Exception)
		{
			// noinspection StringConcatenationArgumentToLogCall
			logger.error("Can't create the item with arguments \"${itemArguments.take(64)}\"", e)
			null
		}
	}

	@JvmStatic
	fun getEnchantment(itemStack: IItemStack?, enchantment: IEnchantment): Int = if (itemStack?.enchantmentTagList == null || itemStack.enchantmentTagList!!.hasNoTags()) 0 else (0 until itemStack.enchantmentTagList!!.tagCount()).map(itemStack.enchantmentTagList!!::getCompoundTagAt).firstOrNull { it.hasKey("ench") && it.getShort("ench").toInt() == enchantment.effectId || it.hasKey("id") && it.getShort("id").toInt() == enchantment.effectId }?.let { tagCompound: INBTTagCompound -> tagCompound.getShort("lvl").toInt() } ?: 0

	@JvmStatic
	fun getEnchantmentCount(itemStack: IItemStack?): Int = if (itemStack?.enchantmentTagList == null || itemStack.enchantmentTagList!!.hasNoTags()) 0 else (0 until itemStack.enchantmentTagList!!.tagCount()).map(itemStack.enchantmentTagList!!::getCompoundTagAt).count { it.hasKey("ench") || it.hasKey("id") }

	@JvmStatic
	fun isStackEmpty(stack: IItemStack?): Boolean = stack == null || classProvider.isItemAir(stack.item)

	@JvmStatic
	private val PATTERN = Pattern.compile("\\d+")
}
